### ISSUE TITLE
Support different Git servers in TeamCityPullRequestInfo

### ISSUE THREAD

**Issue by pascalberger** [cake-build/cake#1618]
### What You Are Seeing?
[TeamCityPullRequestInfo](https://github.com/cake-build/cake/blob/develop/src/Cake.Common/Build/TeamCity/Data/TeamCityPullRequestInfo.cs) currently works only with certain Git servers like Stash, since it expects pull requests in `refs/pull-requests/`. This doesn't work for other Git servers like eg. TFS.

### What is Expected?
It should be possible to use the class with any Git server

### What version of Cake are you using?
0.19.5

### Are you running on a 32 or 64 bit system?
64 bit

### What environment are you running on?  Windows? Linux? Mac?
Windows

### Are you running on a CI Server?  If so, which one?
TeamCity

### How Did You Get This To Happen? (Steps to Reproduce)
* Create a pull request on TFS and trigger the build for the pull request
* `TeamCityPullRequestInfo.IsPullRequest` returns `false`

**Comment by mholo65** [comment]
:+1: e.g. Gerrit uses `refs/changes`, and Github `refs/pull`.

This should also be taken into account in other build systems.

Should the ref be user configurable or should Cake check some known possibilities? I bet Gitlab, Github, Bitbucket, Gerrit, TFS, etc all use different refs for PRs..

**Comment by mvput** [comment]
Sounds best to make this configurable for the user. Setting someting like `TeamCity.VcsProvider = VcsProviders.GitHub`. With that you can put more configuration settings for that particular git host.

**Comment by Kralizek** [comment]
Any information about this?

**Comment by devlead** [comment]
@Kralizek I believe this issue is till up for grabs, I'll tag it so.

**Comment by Kralizek** [comment]
I could try to give a swing at it. Any specific design direction I should consider?

**Comment by devlead** [comment]
@Kralizek cool, then I'll remove the up-for-grabs/help wanted labels.
Read through the [Contribution Guidelines](https://github.com/cake-build/cake/blob/develop/CONTRIBUTING.md), feel free to ping if you have any questions. 
Running release build will catch many code style issues

These are the points that would need to be smarter
https://github.com/cake-build/cake/blob/a0f142af2dfbf18eecc240789529da428cb0d82a/src/Cake.Common/Build/TeamCity/Data/TeamCityPullRequestInfo.cs#L23

https://github.com/cake-build/cake/blob/a0f142af2dfbf18eecc240789529da428cb0d82a/src/Cake.Common/Build/TeamCity/Data/TeamCityPullRequestInfo.cs#L34

Probably 2 private helper methods instead with common smarts to decide which vcs and how to try parse if pull request and its number.

**Comment by Kralizek** [comment]
@devlead do you think a strategy pattern would be overkill? 

I'm thinking of something like:

```csharp
public interface IGitProvider
{
    bool IsPullRequest();

    int? PullRequestNumber();
}

public static class GitProviders
{
     public static readonly IGitProvider GitHub = new GitHubGitProvider();

     public static readonly IGitProvider Stash = new StashGitProvider();

     public static readonly IGitProvider BitBucket = new BitBucketGitProvider();

     ...
}

// usage
TeamCity.GitProvider = GitProviders.GitHub;
...
var isPR = TeamCity.GitProvider.IsPullRequest();
```

Few questions:
1) how to handle non-git repositories?
2) do all git repositories have a pr-number?

Alternatively

```csharp
public interface IPullRequestInfo
{
    bool IsPullRequest { get; }

    int? Number { get; }
}

public interface IGitProvider
{
    IPullRequestInfo GetPullRequest();
}
```

To be used with the Null Object pattern so that the provider never returns `null`.

This way the `IPullRequestInfo` object can be customized to expose provider-specific behaviors, accessible via pattern matching the result of `GetPullRequest`.
```csharp
var pr = TeamCity.GitProvider.GetPullRequest();

if (pr.IsPullRequest && pr as GitHubPullRequest github)
{
    github.DoSomethingSpecialForGitHub();
}
```

**Comment by devlead** [comment]
@Kralizek 

Yes probably overkill initially,  PR / number could probably be inferred from environment variables, Git_Branch pattern in most cases. 

1. I don't know, best consult TeamCity documentation, but I would start with git repos.
2. I don't know, would look at TFS / Azure DevOps, Bitbucket Server, Bitbucket Clpud, Gerrit, etc. own docs.

**Comment by Kralizek** [comment]
By investigating what's available on TeamCity, I found out that there is a Configuration Parameter that contains the URL of the repository: like git@github.com:cake-build/cake.git.

We could use that information to select the proper "parsing logic", granted the user passes that value as an environment variable like with the branch name.

**Comment by devlead** [comment]
Fixed by #2555
